### PR TITLE
Fix case activity field set to allow long details to be exported

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Replacement for https://github.com/civicrm/civicrm-core/pull/17416 which fixed exporting of over-long case activity details field but had unaddressed failing tests & also changed the search class for reasons that I think had evolved to be irrelevant

Before
----------------------------------------
Case exports may fail if they include case activities with details longer than 255 char

After
----------------------------------------
Spec for case activity fields mostly matches non-case activity fields

Technical Details
----------------------------------------
As this removes an IF it is best viewed with https://github.com/civicrm/civicrm-core/pull/17970/files?w=1

As can been seen in the test changes this also fixes some column titles & standardises a few other fields

Comments
----------------------------------------
@demeritcowboy @mattwire

### CODE REVIEW

**Comment by demeritcowboy** [comment]
I'm getting an error and I checked the test site here too and it's the same error. It's not immediately obvious why - maybe a misaligned column somewhere?

`INSERT INTO civicrm_tmp_d_export_9cced9d1a0aec0ffdc260f75b6e5c9d6 (id, c...", "1406 ** Data too long for column 'case_source_contact_id' at row 1"`

```
0 .../CRM/Core/Error.php(148): CRM_Core_Error::backtrace()
1 .../vendor/pear/pear-core-minimal/src/PEAR.php(922): CRM_Core_Error::handle(Object(DB_Error))
2 .../packages/DB.php(998): PEAR_Error->__construct("DB Error: unknown error", -1, 16, (Array:2), "\nINSERT INTO civicrm_tmp_d_export_9cced9d1a0aec0ffdc260f75b6e5c9d6 (`id`, `c...")
3 .../vendor/pear/pear-core-minimal/src/PEAR.php(575): DB_Error->__construct(-1, 16, (Array:2), "\nINSERT INTO civicrm_tmp_d_export_9cced9d1a0aec0ffdc260f75b6e5c9d6 (`id`, `c...")
4 .../vendor/pear/pear-core-minimal/src/PEAR.php(223): PEAR->_raiseError(Object(DB_mysqli), NULL, -1, 16, (Array:2), "\nINSERT INTO civicrm_tmp_d_export_9cced9d1a0aec0ffdc260f75b6e5c9d6 (`id`, `c...", "DB_Error", TRUE)
5 .../packages/DB/common.php(1925): PEAR->__call("raiseError", (Array:7))
6 .../packages/DB/mysqli.php(936): DB_common->raiseError(-1, NULL, NULL, "\nINSERT INTO civicrm_tmp_d_export_9cced9d1a0aec0ffdc260f75b6e5c9d6 (`id`, `c...", "1406 ** Data too long for column 'case_source_contact_id' at row 1")
7 .../packages/DB/mysqli.php(406): DB_mysqli->mysqliRaiseError()
8 .../packages/DB/common.php(1231): DB_mysqli->simpleQuery("\nINSERT INTO civicrm_tmp_d_export_9cced9d1a0aec0ffdc260f75b6e5c9d6 (`id`, `c...")
9 .../packages/DB/DataObject.php(2696): DB_common->query("\nINSERT INTO civicrm_tmp_d_export_9cced9d1a0aec0ffdc260f75b6e5c9d6 (`id`, `c...")
10 .../packages/DB/DataObject.php(1829): DB_DataObject->_query("\nINSERT INTO civicrm_tmp_d_export_9cced9d1a0aec0ffdc260f75b6e5c9d6 (`id`, `c...")
11 .../CRM/Core/DAO.php(439): DB_DataObject->query("\nINSERT INTO civicrm_tmp_d_export_9cced9d1a0aec0ffdc260f75b6e5c9d6 (`id`, `c...")
12 .../CRM/Core/DAO.php(1529): CRM_Core_DAO->query("\nINSERT INTO civicrm_tmp_d_export_9cced9d1a0aec0ffdc260f75b6e5c9d6 (`id`, `c...", TRUE)
13 .../CRM/Export/BAO/Export.php(365): CRM_Core_DAO::executeQuery("\nINSERT INTO civicrm_tmp_d_export_9cced9d1a0aec0ffdc260f75b6e5c9d6 (`id`, `c...")
14 .../CRM/Export/BAO/Export.php(199): CRM_Export_BAO_Export::writeDetailsToTable(Object(CRM_Export_BAO_ExportProcessor), (Array:1), (Array:101))
15 .../CRM/Export/Form/Select.php(386): CRM_Export_BAO_Export::exportComponents(TRUE, (Array:1), (Array:3), "`sort_name` asc", NULL, NULL, 6, " civicrm_case.id IN ( 1 ) ", NULL, 0, 0, (Array:3), NULL)
16 .../CRM/Core/Form.php(504): CRM_Export_Form_Select->postProcess()
17 .../CRM/Core/StateMachine.php(144): CRM_Core_Form->mainProcess()
18 .../CRM/Core/QuickForm/Action/Next.php(43): CRM_Core_StateMachine->perform(Object(CRM_Export_Form_Select_Case), "next", "Next")
19 .../packages/HTML/QuickForm/Controller.php(203): CRM_Core_QuickForm_Action_Next->perform(Object(CRM_Export_Form_Select_Case), "next")
20 .../packages/HTML/QuickForm/Page.php(103): HTML_QuickForm_Controller->handle(Object(CRM_Export_Form_Select_Case), "next")
21 .../CRM/Core/Controller.php(347): HTML_QuickForm_Page->handle("next")
22 .../CRM/Core/Invoke.php(312): CRM_Core_Controller->run((Array:3), NULL)
23 .../CRM/Core/Invoke.php(68): CRM_Core_Invoke::runItem((Array:14))
24 .../CRM/Core/Invoke.php(36): CRM_Core_Invoke::_invoke((Array:3))
25 .../drupal/civicrm.module(459): CRM_Core_Invoke::invoke((Array:3))
26 .../web/includes/menu.inc(527): civicrm_invoke("case", "search")
27 .../web/index.php(21): menu_execute_active_handler()
28 {main}
```

**Comment by demeritcowboy** [comment]
This is the full query. case_source_contact_id is an email address for some reason:

`INSERT INTO civicrm_tmp_d_export_74f676271b91d58422673072469adf79 (id, civicrm_primary_id, contact_type, contact_sub_type, do_not_email, do_not_phone, do_not_mail, do_not_sms, do_not_trade, is_opt_out, legal_identifier, external_identifier, sort_name, display_name, nick_name, legal_name, image_url, preferred_communication_method, preferred_language, preferred_mail_format, hash, contact_source, first_name, middle_name, last_name, prefix_id, suffix_id, formal_title, communication_style_id, email_greeting_id, postal_greeting_id, addressee_id, job_title, gender_id, birth_date, is_deceased, deceased_date, household_name, organization_name, sic_code, user_unique_id, current_employer_id, contact_is_deleted, created_date, modified_date, addressee, email_greeting, postal_greeting, current_employer, location_type, address_id, street_address, street_number, street_number_suffix, street_name, street_unit, supplemental_address_1, supplemental_address_2, supplemental_address_3, city, postal_code_suffix, postal_code, geo_code_1, geo_code_2, manual_geo_code, address_name, master_id, county, state_province, country, phone, phone_ext, phone_type_id, email, on_hold, is_bulkmail, signature_text, signature_html, im_provider, im, openid, world_region, url, contact_id, case_id, case_activity_subject, case_subject, case_status, case_type, case_role, case_deleted, case_recent_activity_date, case_recent_activity_type, case_scheduled_activity_date, case_start_date, case_end_date, case_source_contact_id, case_activity_status, case_activity_duration, case_activity_medium_id, case_activity_details, case_activity_is_auto)
VALUES (1,'325','Individual','','','','','','','','','','User 315, Fake','Fake User 315','','','','','en_US','Both','e9a6a499f0e781e03cd49dbbed0fc991','','Fake','','User 315','','','','Formal','1','1','1','','','1980-04-02','','','','','','','','','2019-04-26 20:15:29','2020-04-02 22:50:01','Fake User 315','Dear Fake','Dear Fake','','','','','','','','','','','','','','','','','','','','','','','444-555-6666','','Phone','fakeemail315@hotmail.com','','','','','','','','','','325','323','Case Subject 315','Case Subject 315','Ongoing','Housing Support','Homeless Services Coordinator is','','2019-01-16 00:00:00','Open Case','2019-01-16 00:00:00','2019-01-16','','foo@example.com','Completed','','Phone','','')`

**Comment by eileenmcnaughton** [comment]
@demeritcowboy thanks - I guess it's converting the name - I'll have to look & see why it's different to on activity export

**Comment by eileenmcnaughton** [comment]
@demeritcowboy I put some hacks in to cope with that field

File: CRM/Activity/BAO/Activity.php [hunk 1]
```diff
@@ -2193,30 +2185,25 @@ public static function exportableFields($name = 'Activity') {
...
           'type' => CRM_Utils_Type::T_BOOLEAN,
         ],
       ];
+      $caseStandardFields = ['activity_subject', 'source_contact_id', 'activity_status', 'activity_duration', 'activity_details'];
+      foreach ($caseStandardFields as $key) {
+        $caseActivityFields['case_' . $key] = $fields[$key];
+      }
+      // This is a hack to get past an inconsistency.
+      // Activity BAO_Query object adds source_contact as a name field but Case BAO_Query squishes it into the id field.
+      $caseStandardFields['source_contact_id']['type'] = CRM_Utils_Type::T_STRING;
```
Comment (demeritcowboy) [inline CRM/Activity/BAO/Activity.php:2203]: Line 2203 needs to be `$caseActivityFields['case_source_contact_id']['type']`.

**Comment by eileenmcnaughton** [comment]
Opps - thanks!

File: CRM/Activity/BAO/Activity.php [hunk 2]
```diff
@@ -2193,30 +2185,25 @@ public static function exportableFields($name = 'Activity') {
...
           'type' => CRM_Utils_Type::T_BOOLEAN,
         ],
       ];
+      $caseStandardFields = ['activity_subject', 'source_contact_id', 'activity_status', 'activity_duration', 'activity_details'];
+      foreach ($caseStandardFields as $key) {
+        $caseActivityFields['case_' . $key] = $fields[$key];
+      }
+      // This is a hack to get past an inconsistency.
+      // Activity BAO_Query object adds source_contact as a name field but Case BAO_Query squishes it into the id field.
+      $caseActivityFields['source_contact_id']['type'] = CRM_Utils_Type::T_STRING;
```
Comment (demeritcowboy) [inline CRM/Activity/BAO/Activity.php:2203]: I should have stated there's two parts of the line that need updating. The whole line needs to be `$caseActivityFields['case_source_contact_id']['type'] = CRM_Utils_Type::T_STRING;`

Comment (eileenmcnaughton) [reply CRM/Activity/BAO/Activity.php]: doh! Serves me for trying to fix it while really working on something else

Comment (demeritcowboy) [reply CRM/Activity/BAO/Activity.php]: No worries.

**Comment by demeritcowboy** [comment]
Have run it. It works except if you do a find cases + export with selected fields.

What happens is
(a) the available activity fields show up multiple times - activity status is even in there thrice, and
(b) it returns way too many records, some of which aren't even case activities.

(There's also a pre-existing E_NOTICE which I already had a lab ticket about.)

This issue seems pretty difficult.

**Comment by eileenmcnaughton** [comment]
@demeritcowboy ouch - I'll take another look - do you have the gitlab on it?

**Comment by demeritcowboy** [comment]
It's here: https://lab.civicrm.org/dev/core/-/issues/1375#note_26818 - see the last paragraph in that comment. I remember thinking I couldn't see a simple way to untangle it.

**Comment by eileenmcnaughton** [comment]
oh wow that's messy. Well getting the phone export thing cleaned up was also crazy messy

**Comment by eileenmcnaughton** [comment]
@demeritcowboy I just tested without this patch & with this patch & get the same....

I created one case of type 'Housing support'

Screen Shot 2020-07-28 at 9 43 08 PM

Screen Shot 2020-07-28 at 9 43 31 PM


Screen Shot 2020-07-28 at 9 47 23 PM

Screen Shot 2020-07-28 at 9 47 40 PM

**Comment by demeritcowboy** [comment]
Does Roland Adams have any non-case activities like Phone Call? Add one with him as the with-contact and then export again.

**Comment by eileenmcnaughton** [comment]
@demeritcowboy hmm no - that wasn't it - same output

Screen Shot 2020-07-29 at 7 38 27 AM

**Comment by demeritcowboy** [comment]
Maybe something about the exact fields selected? I had tried it on the test site here too and was getting the extra activities exported. Let me try the exact selected fields in your screenshot.

**Comment by demeritcowboy** [comment]
There is no activity reporter so I put source contact, but on the test site here I'm getting the extra non-case activities.

**Comment by demeritcowboy** [comment]
I've saved the mapping if you want to see.

**Comment by eileenmcnaughton** [comment]
thanks @demeritcowboy I seem to have replicated now so I'll have to figure out which field triggered it

**Comment by eileenmcnaughton** [comment]
@demeritcowboy OK - the issue is we need to avoid exposing non-case activity fields for selection. I think I have it now - that mapping you saved will be missing the case_ prefix on one or more fields

File: 1 [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+# Your branch is ahead of 'origin/master' by 1 commit.
+#   (use "git push" to publish your local commits)
+#
+# Changes to be committed:
+#	modified:   CRM/Core/BAO/Cache.php
+#	modified:   CRM/Utils/System.php
+#
+# Untracked files:
+#	Civi/Api4/Action/Contact/Merge.php
+#
```
Comment (demeritcowboy) [inline 1:18]: Oops.

But FYI I haven't tried running yet

Comment (eileenmcnaughton) [reply 1]: It has magically disappeared

Comment (demeritcowboy) [reply 1]: What has? I don't see anything. (grin)

File: tests/phpunit/CRM/Export/BAO/ExportTest.php [hunk 4]
```diff
@@ -2288,23 +2288,23 @@ protected function getCaseHeaderDefinition() {
...
       86 => 'Case Status',
       87 => 'Case Type',
       88 => 'Role in Case',
       89 => 'Case is in the Trash',
-      90 => 'case_recent_activity_date',
-      91 => 'case_recent_activity_type',
-      92 => 'case_scheduled_activity_date',
+      90 => 'Activity Actual Date',
+      91 => 'Activity Type',
+      92 => 'Activity Scheduled Date',
```
[the remark this answers is not in the document]
Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Export/BAO/ExportTest.php:2299]: @demeritcowboy so in the export it is now picking up the field titles (from the metadata) before it was using the column_names

[the remark this answers is not in the document]
Comment (demeritcowboy) [reply tests/phpunit/CRM/Export/BAO/ExportTest.php:2299]: Hmm, do you know is that metadata a recent change or has it always been wrong in this export?

[the remark this answers is not in the document]
Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Export/BAO/ExportTest.php:2299]: @demeritcowboy so the change is going from field name to having metadata available & hence using that for the title

[the remark this answers is not in the document]
Comment (demeritcowboy) [reply tests/phpunit/CRM/Export/BAO/ExportTest.php:2299]: Has the metadata always been wrong then? Where do I find this metadata?

[the remark this answers is not in the document]
Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Export/BAO/ExportTest.php:2299]: @demeritcowboy it looks like it is specifically hard-coded here https://github.com/civicrm/civicrm-core/blob/d3e1b01004212f8f749d640ea989c53610928497/CRM/Activity/BAO/Activity.php#L2181

[the remark this answers is not in the document]
Comment (demeritcowboy) [reply tests/phpunit/CRM/Export/BAO/ExportTest.php:2299]: There is no such thing as "actual vs scheduled". There's activity date/time, and then pseudo-fields case_recent_activity_date and case_scheduled_activity_date that are in the db view and refer to the most recent and next scheduled activity on the case. It looks like this has been mixed up for a while. Will come back to it later.

[the remark this answers is not in the document]
Comment (eileenmcnaughton) [reply tests/phpunit/CRM/Export/BAO/ExportTest.php:2299]: @demeritcowboy I'm happy to merge a PR to fix this to whatever you think it should be once this is sorted

**Comment by demeritcowboy** [comment]
I think this is ok for now. But some differences from current:

* When I go to Find Activities, and then export from the results, I see four variations of priority, and two for activity status. But oddly the other activity fields display more like what I would have expected this export to display. Comparing to dmaster.demo I see three variations of activity priority. Also confusing is that dmaster.demo has engagement index listed twice, but here I only see it once. 
* When creating a profile, I also see four variations of activity priority available. And a similar comparison on dmaster.demo as for activity export.

So it's messy both before and after, just different messy.

I also see a couple oddities that are the same before and after, like the preview if you choose Display Name shows different than what actually gets exported. But at this point I can't even remember what the original PR was about.

File: tests/phpunit/CRM/Export/BAO/ExportTest.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (demeritcowboy) [reply tests/phpunit/CRM/Export/BAO/ExportTest.php:2299]: Notes about this: https://lab.civicrm.org/dev/core/-/issues/1916

**Comment by eileenmcnaughton** [comment]
Lol - well the code is a bit more standard :-) And the weirdness level seems similar. Priority is a messy field - the field title should be Priority ID & then the option group name is priority & they would be dismbiguated

If you think it's OK I'll merge

**Comment by demeritcowboy** [comment]
Go for it. I think that test fail is happening regularly now.

**Comment by eileenmcnaughton** [comment]
Yep - test fail is date related..

**Pull request merged**